palette/src/random_sampling/cone.rs

```diff
@@ -0,0 +1,155 @@
+use core::marker::PhantomData;
+
+use crate::float::Float;
+use crate::rgb::RgbSpace;
+use crate::{from_f64, FloatComponent, Hsl, Hsv};
+
+// Based on https://stackoverflow.com/q/4778147 and https://math.stackexchange.com/q/18686,
+// picking A = (0, 0), B = (0, 1), C = (1, 1) gives us:
+//
+// (  sqrt(r1) * r2  ,  sqrt(r1) * (1 - r2) + sqrt(r1) * r2  ) =
+// (  sqrt(r1) * r2  ,  sqrt(r1) - sqrt(r1) * r2 + sqrt(r1) * r2  ) =
+// (  sqrt(r1) * r2  ,  sqrt(r1)  )
+//
+// `sqrt(r1)` gives us the scale of the triangle, `r2` the radius.
+// Substituting, we get `x = scale * radius, y = scale` and thus for cone
+// sampling: `scale = powf(r1, 1.0/3.0)` and `radius = sqrt(r2)`.
+
+pub fn sample_hsv<S, T>(r1: T, r2: T, r3: T) -> Hsv<S, T>
+where
+    T: FloatComponent,
+    S: RgbSpace,
+{
+    let hue = (from_f64::<T>(360.0) * r3).into();
+    let (value, saturation) = (Float::cbrt(r1), Float::sqrt(r2));
+
+    Hsv {
+        hue,
+        saturation,
+        value,
+        space: PhantomData,
+    }
+}
+
+pub fn sample_hsl<S, T>(r1: T, r2: T, r3: T) -> Hsl<S, T>
+where
+    T: FloatComponent,
+    S: RgbSpace,
+{
+    let hue = (from_f64::<T>(360.0) * r3).into();
+    let (saturation, lightness) = if r1 <= from_f64::<T>(0.5) {
+        // Scale it up to [0, 1]
+        let r1 = r1 * from_f64::<T>(2.0);
+        let h = Float::cbrt(r1);
+        let r = h * Float::sqrt(r2);
+        // Scale the lightness back to [0, 0.5]
+        (r / h, h * from_f64::<T>(0.5))
```

Comment:
This is where I mean `h` cancels out. You see how it's multiplied into `r` on line 44 and then divided away again on line 46. The same happens in the other branch.